### ISSUE TITLE
Creating API documentation for onnx_runtime_cpp

### ISSUE THREAD

**Issue by cardboardcode** [xmba15/onnx_runtime_cpp#10]
Hi @xmba15,

My enquiry is: 
## Are you currently working on documenting this codebase's API?

If not, is it okay if I contribute by doing the API documentation?

I find the examples you have included in this repository to be very useful in helping to understand how one can apply **onnxruntime** in CPP. I believe more can benefit from your code better when there is more documentation showing the underlying structures.

**Comment by xmba15** [comment]
@cardboardcode I started this small library to know how onnxruntime works some time ago, and yes, I am not actively working on it recently. I am more than happy if you can help contribute to adding API documentation for it.

**Comment by cardboardcode** [comment]
Okay. 

Will get started on it. No definite date on when this will be completed since I can only do this in my free time. 

Will track the progress here in this thread.

**Comment by cardboardcode** [comment]
## TODO List
The list below shows which files' documentation has been done:
- [ ] ImageClassificationOrtSessionHandler.cpp 
- [ ] ImageRecognitionOrtSessionHandlerBase.cpp
- [ ] ObjectDetectionOrtSessionHandler.cpp
- [ ] OrtSessionHandler.cpp
- [ ] Constants.hpp
- [ ] ImageClassificationOrtSessionHandler.hpp
- [ ] ImageRecognitionOrtSessionHandlerBase.hpp
- [ ] ObjectDetectionOrtSessionHandler.hpp 
- [ ] OrtSessionHandler.hpp 
- [ ] Utility.hpp
- [ ] ort_utility.hpp 

The documentation can be accessed via this link:
Link to be created.

**Comment by cardboardcode** [comment]
Hi @xmba15,

May I clarify what **Imagenet-Trained** `.onnx` model you had in mind for `TestObjectDetection.cpp`?

Asking because, in documenting `TestObjectDetection.cpp`, realized it is not mentioned on the `README.md` and the `.cpp` [only outputs the size of the resulting output tensor](https://github.com/xmba15/onnx_runtime_cpp/blob/35cf4be8ccd59e841eb569800d40c9b504a62e90/examples/TestObjectDetection.cpp#L54).

**Comment by xmba15** [comment]
Hi @cardboardcode Sorry for the late reply.
That TestObjectDetection.cpp is actually one of the earliest examples I wrote for this library. Its main purpose is to test that the API in OrtSessionHandler is giving the right inference output.
Also, at the beginning, I got all the onnx models from the [onnx model zoo](https://github.com/onnx/models/) that I actually didn't know the size of the outputs, so I used TestObjectDetection.cpp to print out to see.

Just in case you are not aware of, I later use this tool to visualize the size of an unknown model downloaded from OSS.
https://netron.app/

Thanks for the documentation work.

**Comment by roachsinai** [comment]
@xmba15 sorry for bother. But after I compile this whole repo, get error:

```
(mp) owin ~/c/G/O/onnx_runtime_cpp master » ./build/Debug/examples/yolov3 data/yolov3/yolov3.onnx data/images/bird_detection.jpg
[/home/xq/code/Git/ONNX/onnx_runtime_cpp/src/OrtSessionHandler.cpp][initSession][Line 204] >>> Model number of inputs: 2

[/home/xq/code/Git/ONNX/onnx_runtime_cpp/src/OrtSessionHandler.cpp][initSession][Line 210] >>> Model number of outputs: 3

zsh: segmentation fault  ./build/Debug/examples/yolov3 data/yolov3/yolov3.onnx
```

I download yolo model from https://github.com/onnx/models/raw/master/vision/object_detection_segmentation/yolov3/model/yolov3-10.onnx

command ` ./build/Debug/examples/TestImageClassification ./data/squeezenet1.1.onnx ./data/images/bird_detection.jpg` run successd, though I don't know why get a lots of verbos:

```
[/home/xq/code/Git/ONNX/onnx_runtime_cpp/src/OrtSessionHandler.cpp][operator()][Line 297] >>> type of input 1: float
794 : shower curtain : 0.068582
922 : menu : 0.0469442
549 : envelope : 0.0451862
619 : lampshade, lamp shade : 0.0370278
991 : coral fungus : 0.0339118

[/home/xq/code/Git/ONNX/onnx_runtime_cpp/src/OrtSessionHandler.cpp][operator()][Line 297] >>> type of input 1: float
794 : shower curtain : 0.068582
922 : menu : 0.0469442
549 : envelope : 0.0451862
619 : lampshade, lamp shade : 0.0370278
991 : coral fungus : 0.0339118

[/home/xq/code/Git/ONNX/onnx_runtime_cpp/src/OrtSessionHandler.cpp][operator()][Line 297] >>> type of input 1: float
794 : shower curtain : 0.068582
922 : menu : 0.0469442
549 : envelope : 0.0451862
619 : lampshade, lamp shade : 0.0370278
991 : coral fungus : 0.0339118

[/home/xq/code/Git/ONNX/onnx_runtime_cpp/src/OrtSessionHandler.cpp][operator()][Line 297] >>> type of input 1: float
794 : shower curtain : 0.068582
922 : menu : 0.0469442
549 : envelope : 0.0451862
619 : lampshade, lamp shade : 0.0370278
991 : coral fungus : 0.0339118
```

---

I thougt the successd command is because the model is squeezenet1.1.onnx which you uploaded in this repo.

And it only have one input, but the model yolov3 I download at the model zoo, the model have 2 inputs:

```
Inputs name:  ['input_1', 'image_shape']
Outputs name:  ['yolonms_layer_1/ExpandDims_1:0', 'yolonms_layer_1/ExpandDims_3:0', 'yolonms_layer_1/concat_2:0']
```

Any suggesstions will be appreciate! Thanks in advace!

**Comment by cardboardcode** [comment]
Hi @xmba15,

The documentation has been tentatively completed. Please view it at the following link:

# https://onnx-runtime-cpp.readthedocs.io/en/latest/index.html

However, I do see the need for **external feedbacks** in order to improve the reading experience of the documentation.

## Enquiry

May I get your feedback on the following documentation aspects and source files?

### 1. Does it seem too wordy on documentations for the examples App `.cpp` files like [Yolov3App](https://onnx-runtime-cpp.readthedocs.io/en/latest/api/examples/Yolov3App.html), [UltraLightFastGenericFaceDetectorApp](https://onnx-runtime-cpp.readthedocs.io/en/latest/api/examples/UltraLightFastGenericFaceDetectorApp.html), [TinyYolov2App](https://onnx-runtime-cpp.readthedocs.io/en/latest/api/examples/TinyYolov2App.html), [TinyYolov2App](https://onnx-runtime-cpp.readthedocs.io/en/latest/api/examples/TinyYolov2App.html), [TestObjectDetection](https://onnx-runtime-cpp.readthedocs.io/en/latest/api/examples/TestObjectDetection.html) and [TestImageClassification](https://onnx-runtime-cpp.readthedocs.io/en/latest/api/examples/TestImageClassification.html)?

### 2. Would a diagrammatic process flow chart be a better alternative, in your opinion?

### 3. Are there source files which are accidentally missed out from your perspective?

### 4. What additional content can be added to better improve the documentations?

**Comment by xmba15** [comment]
Hi @cardboardcode 
First of all, thanks for the hard work (I know it is very time-consuming to read to understand, and document codes that are written by others).

After reading the documentation page, here is my feedback.
1. &2. To be honest, it seems wordy on documentation of the examples. And yes, I think a diagram will give an easier explanation for the users. Just like when I see the diagram in [Codebase Architecture](https://onnx-runtime-cpp.readthedocs.io/en/latest/structure/index.html), I can see the whole structure of the classes.
3. This is a small library and I didn't see source files missed out. However, as you can see from the above comments in this thread, I have made changes in yolov3 sample to use the weights in onnx model zoo. I wonder how these changes can be reflexed back into documentation once it is put into public use.
4. Personally, I would like to see some sample detection result images in the documentation.

Thanks again for this. Can I mention the url to this documentation in the README of this repo?

**Comment by cardboardcode** [comment]
Yup. Please go ahead and mention the url in the `README.md`.

Once again, thank you for this useful library. Will continue to address your feedback once I have time. 

## **TODO**

- [ ] Convert the documentation of Yolov3App to a diagrammatic format.
- [ ] Convert the documentation of UltraLightFastGenericFaceDetectorApp to a diagrammatic format.
- [ ] Convert the documentation of TinyYolov2App to a diagrammatic format.
- [ ] Convert the documentation of TestObjectDetection to a diagrammatic format.
- [ ] Convert the documentation of TestImageClassification to a diagrammatic format.
- [ ] Draft protocol for reflecting up to date documentation that is synchronized with latest commit.
- [ ] Include more sample detection result images in the documentation... (To be elaborated)

**Comment by cardboardcode** [comment]
## TODO 2022
- [ ] Create docs for **YoloX: high-performance anchor-free YOLO by Megvii**
- [ ] Create docs for **Semantic Segmentation Paddle Seg**
